### PR TITLE
Collect external links from various symbol section types

### PR DESCRIPTION
**Pull request by patshaughnessy**
Bug/issue #, if applicable: 124244028

**Note**: This is based on #847; please review and merge that PR first.

## Summary

Today DocC does not resolve external links present inside of various types of symbol documentation sections, such as the deprecation summary, parameters, dictionary keys, or the various sections of REST API documentation (parameters, responses, body  and body parameters). #847 fixes the first two types of symbol sections: deprecation summary and parameters. This follow on PR handles the remaining section types.

## Dependencies

None.

## Testing

1. Set `DOCC_LINK_RESOLVER_EXECUTABLE` to the `test-data-external-resolver` path
2. Set `DOCC_HTML_DIR` to swift-docc-render-artifact checkout path (`/path/to/swift-docc-render-artifact/dist`)
3. In a documentation catalog with some REST API symbols, create a new documentation extension file for one of the REST API symbols, or edit an existing documentation extension file, and add external links to all of the different symbol section types.  For example:

```md
# ``Some-API-Endpoint``

Retrieve some data from our database.

- HTTPParameters:
  - value: For more information about this value, see <doc://com.test.bundle/something> instead.
  - another_value: For more information about this other value, see <doc://com.test.bundle/other> instead.

- HTTPBody: Read this instead: <doc://com.test.bundle/something-else>.

- HTTPBodyParameters:
  - artist: Read more about artists: <doc://com.test.bundle/artists>.
  - userName: Read more about user names: <doc://com.test.bundle/user-names>.

- HTTPResponses:
  - 201: More information here: <doc://com.test.bundle/more-info>.
  - 204: Even more information here: <doc://com.test.bundle/more-info2>.
  - 887: Much more information here: <doc://com.test.bundle/more-info3>.
```

4. In a similar way, in a documentation catalog with some dictionary keys, create a new documentation extension file, or edit an existing documentation extension file, and add external links to the dictionary keys.  For example:

```md
# ``SymbolName``

This is about some symbol.

- DictionaryKeys:
  - key1: The first key has a link: <doc://com.test.bundle/something/related/to/this/key>.
  - key2: The second key also has a link: <doc://com.test.bundle/something/related/to/this/key>.
```

4. Build the documentation
```
swift run docc preview path/to/Something.docc
```

5.  Preview the REST API symbol's page. The external links should be resolved in all of the section types listed above. You should not see an error "No external resolver registered for com.test.bundle."

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ x] Added tests
- [ x] Ran the `./bin/test` script and it succeeded
~~- [ ] Updated documentation if necessary~~

### CODE REVIEW

**Comment by patshaughnessy** [comment]
@swift-ci please test

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift:1167]: The `any` keyword here new in Swift 5.8 so we can't use it yet. 

I've been working to make it _possible_ for us to update to Swift 5.9 but we haven't actually make the update yet. I plan to do that after March 15 when the `release/6.0` branch will be cut from `main`. If this is planned to be merged before then or be cherry-picked into the 6.0 release then it should remove the `any` keyword here.

Comment (d-ronnqvist) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift:1167]: Also, I get the `with...` + closure pattern here but I feel that it mostly results in extra indentation for the caller:

```swift
withExampleDocumentation(
   file1, file2, file3
) { symbol in
   XCTAssertEqual(symbol.something, ...)
}
```

compared to 
```swift
let symbol = withExampleDocumentation(
   file1, file2, file3
)
XCTAssertEqual(symbol.something, ...)

```

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift:1233]: Should this test use `withExampleRESTAPIDocumentation` instead so the documented symbols has a `data` "language"? If we add extra validation in the future to not allow HTTP documentation for Swift/Objective-C types then these tests would fail and it would be the that person who has to go back and fix old tests.

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift:1226]: nit: I don't feel that `forEach` is any more readable than a regular `for-in` loop here (actually, pretty much never but that's a different topic 😉 ).

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift:1368]: Is it / should it be valid to document http responses, parameters, body, etc. for a symbol that doesn't have a `restResponses` `restParameters`, `restBody`,`restBody/parameters`, etc. in the symbol graph file?

If the test data isn't realistic and we add more validation in the future then these tests will fail, without detecting a real bug, and it's the person who adds the validation who has to go back and fix old test data.

**Review by d-ronnqvist — CHANGES_REQUESTED** [review]

File: Sources/SwiftDocC/Semantics/Symbol/Symbol.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Semantics/Symbol/Symbol.swift]: OK now I understand what these methods were for. thanks. I've added a comment explaining this, and also rewrote the code in ExternalReferenceWalker to visit all of the variants of each symbol.

File: Sources/SwiftDocC/Semantics/ExternalLinks/ExternalReferenceWalker.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Semantics/ExternalLinks/ExternalReferenceWalker.swift]: I've replace `visitMarkups` with `visitSection` as you suggested and also added `visitSectionVariants` to avoid some repetition. This uses the `some` keyword so we will have to wait until you can merge #861 to merge this.

File: Sources/SwiftDocC/Semantics/ExternalLinks/ExternalReferenceWalker.swift
Comment (patshaughnessy) [reply Sources/SwiftDocC/Semantics/ExternalLinks/ExternalReferenceWalker.swift]: I wasn't sure either so I ran a benchmark comparison comparing this ("slow"):

```
visitSectionVariants(symbol.abstractSectionVariants)
```

...with this ("fast"):

```
for abstractSection in symbol.abstractSectionVariants.allValues.map(\.variant) {
    for markup in abstractSection.paragraph.children { visitMarkup(markup) }
}
```

There was no significant difference:


```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ slow                 │ fast                 │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ no change¹      │ 0.235 sec            │ 0.236 sec            │
│ Duration for 'convert-total-time'        │ no change²      │ 3.434 sec            │ 3.413 sec            │
│ Duration for 'documentation-processing'  │ no change³      │ 3.135 sec            │ 3.115 sec            │
│ Duration for 'finalize-navigation-index' │ no change⁴      │ 0.016 sec            │ 0.015 sec            │
│ Peak memory footprint                    │ no change⁵      │ 43.6 MB              │ 43.4 MB              │
│ Data subdirectory size                   │ no change⁶      │ 13.5 MB              │ 13.5 MB              │
│ Index subdirectory size                  │ no change⁷      │ 314 KB               │ 314 KB               │
│ Total DocC archive size                  │ no change⁸      │ 17.2 MB              │ 17.2 MB              │
│ Topic Anchor Checksum                    │ no change       │ 508d5249696fb765486f │ 508d5249696fb765486f │
│ Topic Graph Checksum                     │ no change       │ 2b0cd29ff867a49f008a │ 2b0cd29ff867a49f008a │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

So I decided to use the "slow" version since it's easier to read and understand.

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: I refactored all the unit tests anyway, but I did add file and line where needed.

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: This style with a closure is something I'm used to writing Ruby code. Ruby loves to make this sort of code slightly more human readable, at the cost of some indentation.

Anyway I've rewritten the unit tests anyway, and in the new version I didn't use a closure - the helper functions just return values directly.

Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: I also removed the use of `any` here in the test, but we'll need to merge #861 first since I added a use of `some` in the production code in ExternalReferenceWalker.

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: Yes agreed - I've rewritten the unit tests. See the comment below.

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: Ha ha another Ruby idiom. In Ruby we almost always write `some_collection.each do ... end` with a block/closure, and almost never use a for loop, although it is supported. Again the idea is to make the Ruby code a bit more readable.


But since this is Swift I've refactored things to use for-loops :)

File: Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift
Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: Agreed. I found some realistic symbol graphs files that are present under the `Tests/SwiftDocCTests/Test Bundles` directory, and refactored these new unit test to use those instead, while addressing your other comments above. I think the tests are easier to read now, and hopefully will be more reliable.

Comment (patshaughnessy) [reply Tests/SwiftDocCTests/Infrastructure/ExternalReferenceResolverTests.swift]: This is true for the tests related to the HTTP/Rest endpoint symbol, and for the dictionary keys test. I also added a test for normal symbol parameters which was missing earlier. For the symbol parameters I create a fake symbol graph in memory like the other unit tests in this file do.

**Comment by patshaughnessy** [comment]
The CI will likely fail, since we haven't merged #861 yet and this PR relies on that Swift upgrade. I'll run the CI after we merge #861 .

**Review by d-ronnqvist — APPROVED** [review]

**Comment by d-ronnqvist** [comment]
> The CI will likely fail, since we haven't merged #861 yet and this PR relies on that Swift upgrade. I'll run the CI after we merge #861 .

I hadn't planned on cherry-picking #861 into the 6.0 release but maybe I should. I'll talk to the rest of the team to see what they think about it.

**Comment by d-ronnqvist** [comment]
> > The CI will likely fail, since we haven't merged #861 yet and this PR relies on that Swift upgrade. I'll run the CI after we merge #861 .
> 
> I hadn't planned on cherry-picking #861 into the 6.0 release but maybe I should. I'll talk to the rest of the team to see what they think about it.

Update: I will cherry-pick #861 into the 6.0 release but it's currently blocked on the CI (rdar://125077916)

**Comment by patshaughnessy** [comment]
@swift-ci please test

**Pull request merged**
